### ISSUE TITLE
[FEATURE REQUEST] Waterline and "Big data" Query Performance 

### ISSUE THREAD

**Issue by ArdentZeal** [balderdashy/waterline#343]
Hello, we just implemented complex lists into our product and thought about doing some smaller performance tests for searching / querying the objects in them.

We worked with about 700 objects and what we encountered seems pretty detrimental.

While querying for 700 objects happens in under 1 second -->
GET /api/v1/apps/1/schemas/2/objects 200 356ms - 163.92kb

doing the same with just a simple populate (ca. 4 items per object) results in 33000 ms. And this is without performing any postprocessing on the returned objects, just a simple find with *1* populate. 

GET /api/v1/apps/1/schemas/2/objects 200 32996ms - 1.49mb

```
index: function(strucd, modelName, populateParams) {
        return function(req, res, next) {
            var findOptions = ControllerParams.buildFindParams(strucd, req, modelName);
            var query = strucd.models.static[modelName].find(findOptions);
            query.populate('objectproperties');

            query.exec(function afterFind(err, models) {

                if(err) return next(err);

                return res.json(HTTPSTATUS.OK, models);
            });
        }
    },
```

Is there any way to increase that substantially? I would expect 700 objects to be at like 1 second not 33000ms.

If we did anything wrong, call us out :)

Kind Regards

**Comment by dmarcelino** [comment]
What is the status of this? PR #829 was merged and released recently and it improves the speed of `.populate` considerably. @ArdentZeal can you please re-evaluate the performance you are getting?

**Comment by sailsbot** [comment]
Thanks for posting, @ArdentZeal.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tarlepp** [comment]
and again, I think that issues is still open but @sailsbot has closed this one...

**Comment by devinivy** [comment]
Totally open!

**Comment by nbiton** [comment]
Please look at this branch, if you're interested in a similar performance gain, but without breaking the API:
https://github.com/nbiton/waterline/tree/query-optimize

This is, however, way more intrusive to the waterline code than the change that was mentioned at #924

**Comment by sailsbot** [comment]
Thanks for posting, @ArdentZeal.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tarlepp** [comment]
and again, I think that issues is still open but @sailsbot has closed this one...

**Comment by mikermcneil** [comment]
> The list is fully serverside-paginated, so the wire has only ever like 50 to 100 objects to send and the client to show.

From what I understand, the issue is not the server-side pagination, but sort of the equivalent of "database-side pagination" (or potentially the integrator is just slow in the scenario you're running into).  

@tarlepp in the original code for your example [here](https://github.com/tarlepp/Taskboard/commit/f3953dbd4bf518b9d8467407ecbf8c7b95b330a4) the issue is that the call to `.populate()` is not paginated-- so it's actually fetching _all_ associated records.  I took a look at [your models](https://github.com/tarlepp/Taskboard/blob/f3953dbd4bf518b9d8467407ecbf8c7b95b330a4/backend/api/models/User.js#L72) and they seem fine.

Would you mind posting your schema and trying out:

```
History.find({sort: 'createdAt ASC' }).populate('user',{limit: 30}).exec(...)
```

That will limit the returned users for each history record-- if that improves performance, then we know it's a data-coming-in-to-the-pipe issue rather than a crusty-pipe issue.

**Comment by tarlepp** [comment]
Actually each history record just contain one user, so imho that limit should not affect to that at all...

**Comment by mikermcneil** [comment]
@tarlepp k thanks- one more favor, please:  Any difference if you take away `columnName`?
